x-pack/plugins/fleet/server/services/agents/action_status.ts

```diff
@@ -0,0 +1,142 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+import type { ElasticsearchClient } from '@kbn/core/server';
+import pMap from 'p-map';
+
+import { SO_SEARCH_LIMIT } from '../../constants';
+
+import type { FleetServerAgentAction, ActionStatus } from '../../types';
+import { AGENT_ACTIONS_INDEX, AGENT_ACTIONS_RESULTS_INDEX } from '../../../common';
+
+/**
+ * Return current bulk actions
+ */
+export async function getActionStatuses(esClient: ElasticsearchClient): Promise<ActionStatus[]> {
+  let actions = await _getActions(esClient);
+  const cancelledActionIds = await _getCancelledActionId(esClient);
+
+  // Fetch acknowledged result for every action
+  actions = await pMap(
+    actions,
+    async (action) => {
+      const { count } = await esClient.count({
+        index: AGENT_ACTIONS_RESULTS_INDEX,
+        ignore_unavailable: true,
+        query: {
+          bool: {
+            must: [
+              {
+                term: {
+                  action_id: action.actionId,
+                },
+              },
+            ],
+          },
+        },
+      });
```

Comment:
I'm concerned about the number of queries here. We're fetching all actions, then doing a separate query for action results for every action. Instead I think we should try combining that into a single query and use aggregations to get the count value that we need.

Likely something like this (pseudo-ish code):

```ts
let actions = await _getActions(esClient);

let acks = await esClient.search({
  index: AGENT_ACTIONS_RESULTS_INDEX,
  query: {
    bool: {
      // There's some perf/caching advantages to using filter over must
      // See https://www.elastic.co/guide/en/elasticsearch/reference/current/query-filter-context.html#filter-context
      filter: actions.map(a => ({ term: { action_id: a.id } }),
    }
  },
  aggs: {
    ack_counts: { terms: { field: "action_id" } }
  }
})

return actions.map(a => {
  return {
    ...action,
    nbAgentsAck: acks.aggs.ack_counts.buckets.find(b => b.key === a.id).doc_count,
   // other fields
  }
})
```

Reply:
Is that okay if I refactor this in next [pr](https://github.com/elastic/kibana/pull/140428) about Agent activity UI? I already had to make changes to add more info.

Reply:
I'd be okay with refactoring this fetch operation all at once in the next PR. Let's try to land this one and move on to the UI.

Reply:
refactored here https://github.com/elastic/kibana/pull/140510/files#diff-a1a29af3b4d7a9d69539c684d9bbcaf584fd6db8a60eeebfe47da2af803cb621